src/liballoc/arc.rs

```diff
@@ -192,6 +191,44 @@ impl<T> Arc<T> {
         };
         Arc { _ptr: unsafe { NonZero::new(Box::into_raw(x)) } }
     }
+
+    /// Unwraps the contained value if the `Arc<T>` has only one strong reference.
+    /// This will succeed even if there are outstanding weak references.
+    ///
+    /// Otherwise, an `Err` is returned with the same `Arc<T>`.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// #![feature(arc_unique)]
+    /// use std::sync::Arc;
+    ///
+    /// let x = Arc::new(3);
+    /// assert_eq!(Arc::try_unwrap(x), Ok(3));
+    ///
+    /// let x = Arc::new(4);
+    /// let _y = x.clone();
+    /// assert_eq!(Arc::try_unwrap(x), Err(Arc::new(4)));
+    /// ```
+    #[inline]
+    #[unstable(feature = "arc_unique", reason = "needs FCP", issue = "27718")]
+    pub fn try_unwrap(this: Self) -> Result<T, Self> {
+        // See `drop` for why all these atomics are like this
+        if this.inner().strong.compare_and_swap(1, 0, Release) != 1 { return Err(this) }
```

Comment:
I believe this can be replaced with an `Acquire` on the CAS and no fence, because this operation will only succeed if we're the last one out. (Similar to what we do for `make_mut`.) 

- Normally we fence after only if we see we're the last one, but doing the `Acquire` as part of the operation is slightly more efficient in some architectures.
- Since this only succeeds if we're the last strong ref, the `Release` isn't needed; there's no one who would do a corresponding `Acquire`.

(That said, this is fine as-is; you could treat these as optimizations to be done later.)